### PR TITLE
enforce O_CLOEXEC on Windows

### PR DESCRIPTION
**Pull request by fdopen**
1. The `Pervasives.open*` functions, don't open files with `O_CLOEXEC` on Windows (for whatever reason the Windows and *nix implementation differs at this point).
2. Currently all inheritable handles are also inherited to processes created with `Unix.create_process*`. There is no option to disable it.
3. Some windows programs are picky and fail, if a file is already opened by a parallel process (it depends which flags are passed to `CreateFile` in each process).

1)-3) combined with parallelism lead to a mess. Example:

```ocaml
let ch = open_out "test.c" in
output_string ch "int foo(void) { return 0; }";
flush ch;
let _ = Unix.create_process "sleep" [|"sleep";"10"|] Unix.stdin Unix.stdout Unix.stderr in
close_out ch;
let e = Sys.command "cl.exe -c test.c" in
Printf.printf "exit_status:%d\n!" e
```

Although `ch` is closed, when `Sys.command is` called, it's still open inside the `sleep` process and "cl.exe" will likely fail. Similar situations are easily triggered under Windows and compilation will only succeed with `jbuilder -j 1`.

As a workaround `Unix.openfile` is used instead of `Pervasives.open*` and all files are opened with `O_CLOEXEC`. (You have to pass inheritable handles to `Unix.create_process`. But the racy parts are now contained in `go_rec`).

### CODE REVIEW

File: src/future.ml [hunk 1]
```diff
@@ -588,18 +588,27 @@ module Scheduler = struct
...
-        | Fd      fd -> fd
-        | Channel oc -> flush oc; Unix.descr_of_out_channel oc
-      in
-      (fd, Option.some_if tail desc)
+      match desc with
+      | Fd      fd -> (fd, Option.some_if tail desc)
+      | Channel oc ->
+        flush oc;
+        let fd = Unix.descr_of_out_channel oc in
+        if Sys.win32 = false || ocaml_version >= (4,5) then
```
Comment (dra27) [inline src/future.ml]: What's changed in 4.05 that means this doesn't need to be done anymore?

File: src/io.ml [hunk 2]
```diff
@@ -2,11 +2,94 @@ open Import
...
+  let open_out_gen mode perm fln =
+    let fd = open_gen mode perm fln in
+    match Unix.out_channel_of_descr fd with
+    | exception (Unix.Unix_error(e,_,_)) -> eclose e fd fln
+    | oc ->
+      if List.mem Open_text ~set:mode then (
+        try
+          set_binary_mode_out oc false
+        with
+        | x -> close_out oc ; raise x
```
Comment (dra27) [inline src/io.ml]: I'm not totally convinced of the need to handle errors in `set_binary_mode_*`, but it should be `Unix.close fd` if you do

File: src/io.ml [hunk 3]
```diff
@@ -2,11 +2,94 @@ open Import
...
+  let open_in_gen mode perm fln =
+    let fd = open_gen mode perm fln in
+    match Unix.in_channel_of_descr fd with
+    | exception (Unix.Unix_error(e,_,_)) -> eclose e fd fln
+    | ic ->
+      if List.mem Open_text ~set:mode then (
+        try
+          set_binary_mode_in ic false
+        with
+        | x -> close_in ic ; raise x
```
Comment (dra27) [inline src/io.ml]: Similarly, `Unix.close` here

File: src/io.ml [hunk 4]
```diff
@@ -2,11 +2,94 @@ open Import
 
 module P = Pervasives
 
+module W32_io = struct
```
Comment (dra27) [inline src/io.ml]: Just for consistency with other naming, please could this `Win32` rather than `W32`?

**Comment by fdopen** [comment]
Fd leaks?
Both `Pervasives.close_` and `Unix.close` will use `close` once a crt file descriptor is acquired: https://github.com/ocaml/ocaml/blob/782bac2aebfbff848dad4e75b9757151cc2c1476/otherlibs/win32unix/close.c#L31
But only `Pervasives.close_*` will raise the expected exception.

**Comment by dra27** [comment]
My bad - I was under the mistaken impression that having got an fd from a channel that closing the channel released the API handle, not the fd, but I was reading the wrong bit. Sorry!

**Comment by fdopen** [comment]
Ok,
I've changed the style issues and documented the non-obvious use of `Unix.lseek` and different code path for `Unix.dup` on newer OCaml versions.

**Comment by diml** [comment]
@fdopen why not use the `Unix.xxx` functions all the time in `io.ml`? Is it to be conservative? If there is no particular reason, I'd prefer to use the same code everywhere; the fewer code paths the better.

**Comment by rgrinberg** [comment]
Closing as this PR is quite stale. The author or maintainers is welcome to resubmit a more up to date version

**Pull request closed without merging**